File: ecommerce-afnetworking-objc/ecommerce-afnetworking-objc/ViewController.m
//
//  ViewController.m
//  ecommerce-afnetworking-objc
//
//  Created by Mark Prichard on 12/13/15.
//  Copyright © 2015 AppDynamics E2E. All rights reserved.
//

#import "ViewController.h"

#import "AFNetworking.h"

@interface ViewController ()

@end

@implementation ViewController

NSString *sessionId = nil;
NSString *routeId = nil;
NSString *loginThruCheckout = @"Login to Checkout";
int noOfItems = 0;

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view, typically from a nib.
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

- (void)getHttpCookies:(NSURLResponse*)response
                forURL:(NSURL*)url {
    NSDictionary *headers = [(NSHTTPURLResponse*)response allHeaderFields];
    NSArray *cookies =[NSHTTPCookie cookiesWithResponseHeaderFields:headers forURL:url];
    for (int i = 0; i < [cookies count]; i++) {
        NSHTTPCookie *cookie = ((NSHTTPCookie *) [cookies objectAtIndex:i]);
        
        if ([[cookie name] isEqualToString:@"JSESSIONID"]) {
            sessionId = [[cookie properties] objectForKey:NSHTTPCookieValue];
            NSLog(@"Session ID: %@", sessionId);
        }
        if ([[cookie name] isEqualToString:@"ROUTEID"]) {
            routeId = [[cookie properties] objectForKey:NSHTTPCookieValue];
            NSLog(@"RouteID: %@", routeId);
        }
    }
}

- (void)getHttpRequestHeaders:(NSURLRequest*)request {
    NSDictionary *headers = [(NSMutableURLRequest*)request allHTTPHeaderFields];
    NSLog(@"Request Headers: %@", [headers description]);
}

- (void)getHttpResponseHeaders:(NSURLResponse*)response {
    NSDictionary *headers = [(NSHTTPURLResponse*)response allHeaderFields];
    NSLog(@"Response Headers: %@", [headers description]);
}


- (void)doHttpGet:(NSURL*)url {
    NSMutableURLRequest *request = [[AFHTTPRequestSerializer serializer] requestWithMethod:@"GET"
                                                                                 URLString: [url absoluteString]
                                                                                parameters:nil error:nil];
    [request setValue:sessionId forHTTPHeaderField:@"JSESSIONID"];
    [request setValue:routeId forHTTPHeaderField:@"ROUTEID"];
    [request setValue:[[NSUserDefaults standardUserDefaults] objectForKey:@"username"] forHTTPHeaderField:@"USERNAME"];
    [request setValue:@"application/x-www-form-urlencoded; charset=utf-8" forHTTPHeaderField:@"Content-Type"];
    NSLog(@"Request: %@", request);
    [self getHttpRequestHeaders:request];
    
    AFURLSessionManager* manager = [[AFURLSessionManager alloc] initWithSessionConfiguration: [NSURLSessionConfiguration defaultSessionConfiguration]];
    manager.responseSerializer = [AFHTTPResponseSerializer serializer];
    
    [[manager dataTaskWithRequest:request completionHandler:^(NSURLResponse * _Nonnull response, id  _Nullable responseObject, NSError * _Nullable error) {
        
        if (!error) {
            NSLog(@"Response: %@", response);
        } else {
            NSLog(@"Error: %@", response);
        }
    }] resume];
    
}

- (IBAction)loginClicked:(id)sender {
    NSString *baseURL = [[NSUserDefaults standardUserDefaults] objectForKey:@"url"];
    NSString *relativeURL = @"/rest/user/login";
    
    NSString *postString = @"username=";
    postString = [postString stringByAppendingString: [[NSUserDefaults standardUserDefaults] objectForKey:@"username"]];
    postString = [postString stringByAppendingString:@"&password="];
    postString = [postString stringByAppendingString:[[NSUserDefaults standardUserDefaults] objectForKey:@"password"]];
    
    NSData *postBody = [postString dataUsingEncoding:NSUTF8StringEncoding];
    NSLog(@"POST body: %@", [[NSString alloc] initWithData:postBody encoding:NSUTF8StringEncoding]);

    NSMutableURLRequest *request = [[AFHTTPRequestSerializer serializer] requestWithMethod:@"POST"
                                                                             URLString: [baseURL stringByAppendingString:relativeURL]
                                                                            parameters:nil error:nil];
    [request setValue:@"application/x-www-form-urlencoded; charset=utf-8" forHTTPHeaderField:@"Content-Type"];
    [request setHTTPBody:postBody];
    NSLog(@"Request: %@", request);
    [self getHttpRequestHeaders:request];
    
    AFURLSessionManager* manager = [[AFURLSessionManager alloc] initWithSessionConfiguration: [NSURLSessionConfiguration defaultSessionConfiguration]];
    manager.responseSerializer = [AFHTTPResponseSerializer serializer];
     
    [[manager dataTaskWithRequest:request completionHandler:^(NSURLResponse * _Nonnull response, id  _Nullable responseObject, NSError * _Nullable error) {
        
        if (!error) {
            NSLog(@"Response: %@", response);
            [self getHttpCookies:response forURL:[NSURL URLWithString:[baseURL stringByAppendingString:relativeURL]]];
        } else {
            NSLog(@"Error: %@", error);
        }
    }] resume];

}

- (IBAction)getItemsClicked:(id)sender {
    NSString *baseURL = [[NSUserDefaults standardUserDefaults] objectForKey:@"url"];
    NSString *relativeURL = @"/rest/items/all";

    NSURL *url = [NSURL URLWithString:[baseURL stringByAppendingString:relativeURL]];
    [self doHttpGet:url];
}

- (IBAction)addToCartClicked:(id)sender {
    int x = arc4random() % 10;
    noOfItems++;
    
    NSString *baseURL = [[NSUserDefaults standardUserDefaults] objectForKey:@"url"];
    NSString *relativeURL = [NSString stringWithFormat:@"/rest/cart/%d", x];
    
    NSURL *url = [NSURL URLWithString: [baseURL stringByAppendingString:relativeURL]];
    [self doHttpGet:url];
    
}

- (IBAction)checkoutClicked:(id)sender {
    NSString *baseURL = [[NSUserDefaults standardUserDefaults] objectForKey:@"url"];
    NSString *relativeURL = @"/rest/cart/co";
    
    NSURL *url = [NSURL URLWithString: [baseURL stringByAppendingString:relativeURL]];
    [self doHttpGet:url];
}

- (IBAction)settingsClicked:(id)sender {
    NSUserDefaults *standardDefaults = [NSUserDefaults standardUserDefaults];
    
    NSLog(@"ECommerce URL: %@", [standardDefaults objectForKey:@"url"]);
    NSLog(@"EUM Collector: %@", [standardDefaults objectForKey:@"collectorUrl"]);
    NSLog(@"EUM Key: %@", [standardDefaults objectForKey:@"appKey"]);
    NSLog(@"Username: %@", [standardDefaults objectForKey:@"username"]);
    NSLog(@"Password: %@", [standardDefaults objectForKey:@"password"]);
}

@end

File: ecommerce-ios-swift/ecommerce-ios-swift/ViewController.swift
//
//  ViewController.swift
//  ecommerce-ios-swift
//
//  Created by Mark Prichard on 11/16/15.
//  Copyright © 2015 AppDynamics. All rights reserved.
//

import UIKit

class ViewController: UIViewController {
    var jsessionid: String = String()
    var routeid: String = String()
    var noOfItems: Int64 = 0
    let loginThruCheckout: String = "Login to Checkout"

    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view, typically from a nib.
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
    func getHttpCookies(response: NSURLResponse,
                            url : NSURL) {
        let infoPoint = ADEumInstrumentation.beginCall(self, selector: __FUNCTION__);
        let httpResponse = response as! NSHTTPURLResponse
        let headers = httpResponse.allHeaderFields as? [String : String]
        let cookies = NSHTTPCookie.cookiesWithResponseHeaderFields(headers!, forURL: httpResponse.URL!)
        
        for cookie in cookies {
            if cookie.name == "JSESSIONID" {
                jsessionid = cookie.value
            }
            if cookie.name == "ROUTEID" {
                routeid = cookie.value
            }
        }
        ADEumInstrumentation.endCall(infoPoint)
    }
    
    func getHttpRequestHeaders(request: NSMutableURLRequest) {
        let infoPoint = ADEumInstrumentation.beginCall(self, selector: __FUNCTION__);
        let httpRequest = request
        let headers = httpRequest.allHTTPHeaderFields! as [String : String]
        NSLog("HTTP Request Headers: " + headers.description)
        ADEumInstrumentation.endCall(infoPoint)
    }

    func doHttpPost(url: NSURL) {
        let infoPoint = ADEumInstrumentation.beginCall(self, selector: __FUNCTION__);
        let request = NSMutableURLRequest(URL: url)
        let session = NSURLSession.sharedSession()

        let username = NSUserDefaults.standardUserDefaults().stringForKey("username")
        let password = NSUserDefaults.standardUserDefaults().stringForKey("password")
        let postString = "username=" + username! + "&password=" + password!
        let postBody = (postString as NSString).dataUsingEncoding(NSUTF8StringEncoding)

        request.HTTPBody = postBody
        request.HTTPMethod = "POST"
        request.addValue("application/x-www-form-urlencoded; charset=utf-8", forHTTPHeaderField: "Content-Type")

        NSLog(request.description)
        getHttpRequestHeaders(request)
        NSLog("HTTP Request Body: " + postString)
        
        session.dataTaskWithRequest(request, completionHandler: {data, response, error -> Void in
            if let httpResponse = response as? NSHTTPURLResponse {
                self.getHttpCookies(httpResponse, url: url)
                NSLog(httpResponse.description)
            }
            if (error != nil) {
                print("Error: \(error)")
                return
            }
        }).resume()
        ADEumInstrumentation.endCall(infoPoint)
    }

    func doHttpGet(url: NSURL) {
        let infoPoint = ADEumInstrumentation.beginCall(self, selector: __FUNCTION__);
        let request = NSMutableURLRequest(URL: url)
        let session = NSURLSession.sharedSession()
        
        request.HTTPMethod = "GET"
        request.setValue(jsessionid, forHTTPHeaderField: "JSESSIONID")
        request.setValue(NSUserDefaults.standardUserDefaults().stringForKey("username")!, forHTTPHeaderField: "USERNAME")
        request.setValue(routeid, forHTTPHeaderField: "ROUTEID")
        NSLog(request.description)
        getHttpRequestHeaders(request)
        
        session.dataTaskWithRequest(request, completionHandler: {(data, response, error) in
            if let httpResponse = response as? NSHTTPURLResponse {
                NSLog(httpResponse.description)
                if (data != nil) {
                    //print(NSString(data: (data)!, encoding: NSUTF8StringEncoding)!)
                }
            }
            if (error != nil) {
                print("Error: \(error)")
                return
            }
        }).resume()
        ADEumInstrumentation.endCall(infoPoint)
    }
    
    @IBAction func loginClicked(sender: AnyObject) {
        let baseUrl = NSUserDefaults.standardUserDefaults().stringForKey("url")
        let relativeUrl = "/rest/user/login"
        let url = baseUrl! + relativeUrl
        
        ADEumInstrumentation.leaveBreadcrumb("Login")
        ADEumInstrumentation.setUserData("Username", value: NSUserDefaults.standardUserDefaults().stringForKey("username")!, persist: false)
        ADEumInstrumentation.setUserData("Password", value: NSUserDefaults.standardUserDefaults().stringForKey("password")!, persist: false)
        ADEumInstrumentation.startTimerWithName(loginThruCheckout)

        self.doHttpPost(NSURL(string: url)!)
    }

    @IBAction func getItemsClicked(sender: AnyObject) {
        let baseUrl = NSUserDefaults.standardUserDefaults().stringForKey("url")
        let relativeUrl = "/rest/items/all"
        let url = baseUrl! + relativeUrl
        
        ADEumInstrumentation.leaveBreadcrumb("getItems")
        
        self.doHttpGet(NSURL(string: url)!)
    }
    
    @IBAction func addToCartClicked(sender: AnyObject) {
        let baseUrl = NSUserDefaults.standardUserDefaults().stringForKey("url")
        let relativeUrl = "/rest/cart/" + String(arc4random_uniform(10))
        let url = baseUrl! + relativeUrl
        
        ADEumInstrumentation.leaveBreadcrumb("addToCart")
        noOfItems++

        self.doHttpGet(NSURL(string: url)!)
    }
    
    @IBAction func checkoutClicked(sender: AnyObject) {
        let baseUrl = NSUserDefaults.standardUserDefaults().stringForKey("url")
        let relativeUrl = "/rest/cart/co"
        let url = baseUrl! + relativeUrl
        
        ADEumInstrumentation.leaveBreadcrumb("Checkout")
        ADEumInstrumentation.reportMetricWithName("Number of Items", value: noOfItems)

        self.doHttpGet(NSURL(string: url)!)
        ADEumInstrumentation.stopTimerWithName(loginThruCheckout)
    }
    
    @IBAction func settingsClicked(sender: AnyObject) {
        print("ECommerce URL: " + NSUserDefaults.standardUserDefaults().stringForKey("url")!)
        print("Collector URL: " + NSUserDefaults.standardUserDefaults().stringForKey("collectorUrl")!)
        print("App Key: " + NSUserDefaults.standardUserDefaults().stringForKey("appKey")!)
        print("Username: " + NSUserDefaults.standardUserDefaults().stringForKey("username")!)
        print("Password: " + NSUserDefaults.standardUserDefaults().stringForKey("password")!)
    }
}
